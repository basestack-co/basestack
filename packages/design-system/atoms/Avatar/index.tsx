import React, { memo } from "react";
import { SpaceProps } from "styled-system";
import { Container, Image } from "./styles";

export type Size = "large" | "medium" | "small";

interface AvatarProps extends SpaceProps {
  /**
   * Change avatar size
   */
  size?: Size;
  /**
   * Image src
   */
  src?: string;
  /**
   * User name for initials
   */
  userName?: string;
  /**
   * Image Alt
   */
  alt: string;
}

const handleSize = (size: Size) => {
  switch (size) {
    case "large":
      return 56;
    default:
    case "medium":
      return 40;
    case "small":
      return 24;
  }
};

const Avatar = ({
  src,
  size = "medium",
  userName = "user name",
  alt,
  ...props
}: AvatarProps) => {
  const imageSize = handleSize(size);

  return (
    <Container data-testid="avatar" size={imageSize} {...props}>
      <Image
        alt={alt}
        src={
          src ||
          `https://avatars.dicebear.com/api/initials/${userName}.svg?size=${imageSize}&fontSize=40`
        }
      />
    </Container>
  );
};

export default memo(Avatar);
