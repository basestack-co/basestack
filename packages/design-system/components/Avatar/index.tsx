import { memo } from "react";
import { SpaceProps } from "styled-system";
import { Container, Image } from "./styles";

export type Size = "large" | "medium" | "small" | "xSmall";

export interface AvatarProps extends SpaceProps {
  /**
   * Change avatar size
   */
  size?: Size;
  /**
   * Image src
   */
  src?: string;
  /**
   * User name for initials
   */
  userName?: string;
  /**
   * Image Alt
   */
  alt: string;
  /**
   * Changes the shape from round to squared
   */
  round?: boolean;
}

const handleSize = (size: Size) => {
  switch (size) {
    case "large":
      return 56;
    default:
    case "medium":
      return 40;
    case "small":
      return 36;
    case "xSmall":
      return 28;
  }
};

const Avatar = ({
  src,
  size = "medium",
  userName = "user name",
  round = true,
  alt,
  ...props
}: AvatarProps) => {
  const imageSize = handleSize(size);

  return (
    <Container round={round} data-testid="avatar" size={imageSize} {...props}>
      <Image
        alt={alt}
        src={
          src ||
          `https://avatars.dicebear.com/api/initials/${userName}.svg?size=${imageSize}&fontSize=40`
        }
      />
    </Container>
  );
};

export default memo(Avatar);
