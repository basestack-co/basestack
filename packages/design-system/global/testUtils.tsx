import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
