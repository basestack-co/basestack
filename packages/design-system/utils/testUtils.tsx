import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
