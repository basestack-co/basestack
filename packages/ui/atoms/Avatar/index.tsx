import React from "react";
import styled, { css } from "styled-components";
import { space, SpaceProps } from "styled-system";

export type Size =
  | "xxLarge"
  | "xLarge"
  | "large"
  | "medium"
  | "small"
  | "xSmall";

interface AvatarProps extends SpaceProps {
  /**
   * Change avatar size
   */
  size?: Size;
  /**
   * Image src
   */
  src?: string;
  /**
   * User name for initials
   */
  userName?: string;
  /**
   * Image Alt
   */
  alt: string;
}

const handleTextVariants = (size?: Size) => {
  switch (size) {
    case "xxLarge":
      return css`
        height: 80px;
        width: 80px;
      `;
    case "xLarge":
      return css`
        height: 62px;
        width: 62px;
      `;
    case "large":
      return css`
        height: 56px;
        width: 56px;
      `;
    default:
    case "medium":
      return css`
        height: 40px;
        width: 40px;
      `;
    case "small":
      return css`
        height: 30px;
        width: 30px;
      `;
    case "xSmall":
      return css`
        height: 24px;
        width: 24px;
      `;
  }
};

const handleInitialsSize = (size?: Size) => {
  switch (size) {
    case "xxLarge":
      return 80;
    case "xLarge":
      return 62;
    case "large":
      return 56;
    default:
    case "medium":
      return 40;
    case "small":
      return 30;
    case "xSmall":
      return 24;
  }
};

const Container = styled.div<{ size?: Size }>`
  ${({ size }) => handleTextVariants(size)};
  border-radius: 50%;
  overflow: hidden;
  flex-shrink: 0;
  ${space};
`;

const Image = styled.img`
  height: 100%;
  width: 100%;
  object-fit: cover;
`;

const Avatar = ({
  src,
  size,
  userName = "user name",
  alt,
  ...props
}: AvatarProps) => {
  const initialsSize = handleInitialsSize(size);

  return (
    <Container size={size} {...props}>
      <Image
        alt={alt}
        src={
          src ||
          `https://avatars.dicebear.com/api/initials/${userName}.svg?size=${initialsSize}&fontSize=40`
        }
      />
    </Container>
  );
};

export default Avatar;
