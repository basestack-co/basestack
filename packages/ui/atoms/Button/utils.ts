import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
