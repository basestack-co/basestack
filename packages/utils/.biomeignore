# Node dependencies
node_modules

# Next.js build output
.next
out

# Environment files
.env
.env.*
!.env.example

# Logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
pnpm-debug.log*

# OS / editor files
.DS_Store
Thumbs.db

# Package manager cache
.pnp
.pnp.js
.pnp.cjs

# Test coverage
coverage

# Build artifacts
dist
build

# Temporary files
.tmp
temp