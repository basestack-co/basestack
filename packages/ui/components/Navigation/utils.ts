// Types
import { Translate } from "next-translate";
import { NextRouter } from "next/router";
// Utils
import { config } from "@basestack/utils";
import { signOut } from "next-auth/react";

export const getInternalLinks = (t: Translate, formId: string) => {
  return [
    {
      icon: "flag",
      activeText: "Submissions",
      to: `/form/${formId}/submissions`,
      text: t("internal.submissions"),
    },
    {
      icon: "flag",
      activeText: "Setup",
      to: `/form/${formId}/setup`,
      text: t("internal.setup"),
    },
    {
      icon: "settings",
      activeText: "Settings",
      to: `/form/${formId}/settings/general`,
      text: t("internal.settings"),
    },
  ];
};

export const getExternalLinks = (t: Translate) => {
  return [
    {
      icon: "description",
      activeText: "Documentation",
      to: config.urls.docs.flags.base,
      isExternal: true,
      text: t("external.docs"),
    },
    {
      icon: "link",
      activeText: "Github",
      to: config.urls.repo,
      isExternal: true,
      text: t("external.github"),
    },
  ];
};

export const getAvatarDropdownList = (
  t: Translate,
  router: NextRouter,
  onCreateForm: () => void,
) => {
  return [
    {
      icon: "add_circle",
      text: t("create.form"),
      onClick: onCreateForm,
      separator: true,
    },
    {
      icon: "settings",
      text: t("dropdown.settings"),
      onClick: () =>
        router.push({
          pathname: "/user/profile/settings",
        }),
    },
    {
      icon: "logout",
      text: t("dropdown.logout"),
      onClick: signOut,
    },
  ];
};
