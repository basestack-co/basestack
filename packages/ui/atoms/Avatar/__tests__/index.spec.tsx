import React from "react";
import { cleanup } from "@testing-library/react";
import "jest-styled-components";
import { renderWithTheme } from "../../../utils/testUtils";
import Avatar from "..";

describe("Avatar Atom tests", () => {
  afterEach(cleanup);

  test("render Avatar with default props", () => {
    const { asFragment } = renderWithTheme(
      <Avatar
        src="https://images.pexels.com/photos/3586798/pexels-photo-3586798.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500"
        alt="user_avatar"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("render Avatar with user name initials", () => {
    const { asFragment } = renderWithTheme(
      <Avatar userName="FlÃ¡vio Amaral" alt="user_avatar" />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
