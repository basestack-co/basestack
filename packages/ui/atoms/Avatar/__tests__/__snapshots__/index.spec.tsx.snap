// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Atom tests render Avatar with default props 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL bZDTjr"
  >
    <img
      alt="user_avatar"
      class="sc-gsDKAQ jPeROq"
      src="https://images.pexels.com/photos/3586798/pexels-photo-3586798.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500"
    />
  </div>
</DocumentFragment>
`;

exports[`Avatar Atom tests render Avatar with user name initials 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL bZDTjr"
  >
    <img
      alt="user_avatar"
      class="sc-gsDKAQ jPeROq"
      src="https://avatars.dicebear.com/api/initials/FlÃ¡vio Amaral.svg?size=40&fontSize=40"
    />
  </div>
</DocumentFragment>
`;
